flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/session/FunctionCreator.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.gateway.api.session;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.configuration.ReadableConfig;
+import org.apache.flink.table.catalog.Catalog;
+import org.apache.flink.table.module.Module;
+
+import java.util.Map;
+
+/**
+ * An interface used to create {@link Module} or {@link Catalog}.
+ *
+ * @param <R> type of the return value
+ */
+@PublicEvolving
+public interface FunctionCreator<R> {
+
+    /**
+     * Applies this function to the given arguments.
+     *
+     * @return the function result
+     */
+    R create(Map<String, String> options, ReadableConfig configuration, ClassLoader classLoader);
```

Comment:
I think we should not let the SqlGatewayService determine the options. It's the `Endpoint`'s responsibility to determine the options.